homeassistant/components/p1_monitor/__init__.py

```diff
@@ -0,0 +1,90 @@
+"""The P1 Monitor integration."""
+from __future__ import annotations
+
+from typing import TypedDict
+
+from p1monitor import P1Monitor, Phases, Settings, SmartMeter
+
+from homeassistant.components.sensor import DOMAIN as SENSOR_DOMAIN
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_HOST
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+from .const import (
+    DOMAIN,
+    LOGGER,
+    SCAN_INTERVAL,
+    SERVICE_PHASES,
+    SERVICE_SETTINGS,
+    SERVICE_SMARTMETER,
+)
+
+PLATFORMS = (SENSOR_DOMAIN,)
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up P1 Monitor from a config entry."""
+
+    coordinator = P1MonitorDataUpdateCoordinator(hass, entry=entry)
+    try:
+        await coordinator.async_config_entry_first_refresh()
+    except ConfigEntryNotReady:
+        await coordinator.p1monitor.close()
+        raise
+
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][entry.entry_id] = coordinator
+
+    hass.config_entries.async_setup_platforms(entry, PLATFORMS)
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload P1 Monitor config entry."""
+    unload_ok = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
+    if unload_ok:
+        coordinator = hass.data[DOMAIN].pop(entry.entry_id)
+        await coordinator.p1monitor.close()
+    return unload_ok
+
+
+class P1MonitorData(TypedDict):
+    """Class for defining data in dict."""
+
+    SERVICE_SMARTMETER: SmartMeter
+    SERVICE_PHASES: Phases
+    SERVICE_SETTINGS: Settings
+
+
+class P1MonitorDataUpdateCoordinator(DataUpdateCoordinator[P1MonitorData]):
+    """Class to manage fetching P1 Monitor data from single endpoint."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        *,
+        entry: ConfigEntry,
```

Comment:
This is not needed, `DataUpdateCoordinator` already get this and set it to `self.config_entry`

Reply:
If I remove this I'll break some code :(

Reply:
Move the super call up to be the first